clean up ravendb cluster page (#5355)
Particular/docs.particular.net: persistence/ravendb/cluster-configuration.md

versions: '[6.0,)'
---

DANGER: Running in cluster-configurations without the use of cluster-wide transactions for business data could result in data loss that is impossible to detect. Due to the risk data loss, RavenDB clusters are currently not supported.

> RavenDB version 4 and higher supports configurations with multiple nodes. Support for configurations with a single leader and multiple watchers was originally introduced under the assumption that it’s safe to run without cluster-wide transactions. However, it was discovered that all databases on RavenDB nodes within a cluster behave as primary nodes. That means that any database in a cluster can accept writes. Committing any data safely requires that the data be successfully persisted across a majority of the nodes within the cluster. Optimistic concurrency control cannot be used in cluster scenarios and the only way to enforce this is by using cluster-wide transactions.

Reviewer: What about something like this:

```suggestion
RavenDB version 4 and higher supports cluster configurations with multiple cluster nodes. Database nodes of a database group distributed across multiple cluster nodes are all operating in master mode, that means that any database node in a cluster can accept writes. Due to missing optimistic concurrency control for cluster-wide transactions, unintended data loss might happen on concurrent writes to different database nodes. Therefore, NServiceBus.RavenDB currently does not support cluster configurations yet.
```

Author: Committed a slightly modified version of the suggestion.

versions: '[6.0,)'
---

> DANGER: Running in cluster-configurations without the use of cluster-wide transactions for business data could result in data loss that is impossible to detect. Due to the risk data loss, RavenDB clusters are currently not supported.

Reviewer 2: ```suggestion
DANGER: Cluster configurations and cluster-wide transactions are currently not supported in the RavenDB persistence. 
```

> RavenDB version 4 and higher supports multi-node cluster configurations. Database nodes in a database group, distributed across multiple cluster nodes are all operating in master mode, that means that any database node in the cluster can accept writes. Due to missing optimistic concurrency control for cluster-wide transactions, data loss may happen on concurrent writes to different database nodes. Due to this problem, NServiceBus.RavenDB does not currently support cluster configurations.

Reviewer 2: ```suggestion
RavenDB version 4 and higher supports multi-node cluster configurations. The RavenDB persistence will guard against cluster configurations, as they can lead to data loss if not used correctly (with cluster-wide transactions).

Database nodes in a database group, distributed across multiple cluster nodes are all operating in master mode, which means that any database node in the cluster can accept writes. Due to missing optimistic concurrency control for cluster-wide transactions, data loss may happen on concurrent writes to different database nodes. 
```

> DANGER: Not using cluster-wide transactions in the current form for business data could result in data loss that is impossible to detect. Due to the risk data loss, RavenDB clusters are currently not supported.

Reviewer 2: ```suggestion
DANGER: Running in cluster-configurations without the use of cluster-wide transactions for business data could result in data loss that is impossible to detect. Due to the risk data loss, RavenDB clusters are currently not supported.
```

> DANGER: Not using cluster-wide transactions in the correct form for the business data could result in data loss that is impossible to detect. Due to the potential for data loss, support for RavenDB clusters was withdrawn.

Reviewer: ```suggestion
DANGER: Not using cluster-wide transactions in the current form for business data could result in data loss that is impossible to detect. Due to the risk data loss, RavenDB clusters are currently not supported.
```

we really need to avoid the "withdrawn" wording imo

> Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that on their business data.

Reviewer 2: ```suggestion
```

> Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that across the business data and the NServiceBus data.

Author: @lailabougria how does this sound?

```suggestion
Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that on their business data.
```

## How to move forward

> - If the RavenDB server is configured in a cluster, move back to using a single node. This may not be ideal, as it introduces a single point of failure in the system.

Reviewer: ```suggestion
- If the database is configured as a multi-node database group, move back to using a single database node by removing additional nodes.
```

versions: '[6.0,)'
---

DANGER: Cluster configurations and cluster-wide transactions are currently not supported in the RavenDB persistence. 

RavenDB version 4 and higher supports multi-node cluster configurations. The RavenDB persistence will guard against cluster configurations, as they can lead to data loss if not used correctly (with cluster-wide transactions).

Database nodes in a database group, distributed across multiple cluster nodes are all operating in master mode, which means that any database node in the cluster can accept writes. Due to missing optimistic concurrency control for cluster-wide transactions, data loss may happen on concurrent writes to different database nodes. 

> If the database is configured as a multi-node cluster, remove all nodes but one. 

Reviewer: I assume this should be: 

```suggestion
include: config
```

> DANGER: Make sure the node is empty before doing so, removing a node that contains a database will remove the database.

Reviewer: Covered by the partial

```suggestion
```

RavenDB version 4 and higher supports configurations with multiple nodes. Support for configurations with a single leader and multiple watchers was originally introduced under the assumption that it’s safe to run without cluster-wide transactions. However, it was discovered that all databases on RavenDB nodes within a cluster behave as primary nodes. That means that any database in a cluster can accept writes. Committing any data safely requires that the data be successfully persisted across a majority of the nodes within the cluster. Optimistic concurrency control cannot be used in cluster scenarios and the only way to enforce this is by using cluster-wide transactions.

> Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that across the business data and the NServiceBus data.

Reviewer 2: This sounds off, it's not up to the user to implement it for NServiceBus data, only for business data. We would take care of the NServiceBus part.

Author: @lailabougria how does this sound?

```suggestion
Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that on their business data.
```

Using cluster-wide transactions correctly is complex. Cluster-wide transactions make use of compare-exchange values in order to guarantee consistent writes across the majority of nodes in the cluster. With the current RavenDB api, it’s up to the user to properly implement that across the business data and the NServiceBus data.

> Using cluster-wide transactions correctly requires all the following:

Reviewer: this whole part feels off. It is too much on the explanation side to be at the beginning of the document. Should we just start with a simple "cluster configurations are not supported. The End." And then maybe explain in a dedicated section why, for users that are more interested on the topic?

Author: I moved the red box that says it is unsupported to the top, see if that addresses your concern?

> DANGER: Not using cluster-wide transactions in the correct form for the business data could result in data loss that is impossible to detect. Due to the potential for data loss, support for RavenDB clusters was withdrawn.

Reviewer 2: ```suggestion
DANGER: Running on cluster configurations without the use of cluster-wide transactions for business data could result in data loss that is impossible to detect. Due to the potential for data loss, support for RavenDB clusters was withdrawn.
```

Author: This has changed, see if the current is any better?

## How to move forward

> - If the RavenDB server is configured in a cluster, move back to using a single node. This may not be ideal, as it introduces a single point of failure in the system.

Reviewer 2: ```suggestion
- If the RavenDB server is configured in a cluster, move back to using a single database node. This may not be ideal, as it introduces a single point of failure in the system.
```

Author: This is included in Tim's suggestion which I've incorporated.

- Move to another persistence. This would require migrating data away from RavenDB to [another persistence](/persistence).

> All customers currently using cluster-wide transactions are encouraged to [contact support](https://particular.net/support) to discuss options for transitioning into a safe configuration that causes no data loss.

Reviewer: that seems to be part communication that might go into the release notes for the hotfix and/or emails going out, but doesn't need to be on the documentation?

Author: Agreed, I'll remove it from here.
